File: apps/windows/src/link/SecureLink.cpp
#include "SecureLink.h"

#include "../crypto/Identity.h"

namespace tether::link {

static std::string shortId(const std::string& id) { return id.substr(0, 8) + "..."; }

SecureLink::SecureLink(Options opts) : opts_(std::move(opts)) {
    deviceId_ = tether::identity::deviceIdFromPublicKey(opts_.staticKeypair.publicKey);
    remoteId_ = opts_.peerDeviceId;

    tether::signaling::SignalingClient::Callbacks cbs;
    cbs.onRegistered = [this](int64_t hb) { onRegistered(hb); };
    cbs.onDeliver = [this](const std::string& from, const Bytes& payload) { onDeliver(from, payload); };
    cbs.onPeerStatus = [](const std::string&, bool) {};
    cbs.onError = [this](const std::string& code, const std::string& msg) { onError(code, msg); };

    signaling_ = std::make_unique<tether::signaling::SignalingClient>(
        opts_.serverUrl, deviceId_, opts_.staticKeypair.publicKey, std::move(cbs));
}

void SecureLink::log(const std::string& line) {
    if (opts_.logger) opts_.logger(line);
}

std::future<void> SecureLink::connect() {
    auto fut = registered_.get_future();
    signaling_->Connect();
    return fut;
}

std::future<void> SecureLink::pair() {
    auto fut = paired_.get_future();
    if (opts_.role == Role::Initiator) {
        if (opts_.peerStatic.empty() || remoteId_.empty()) {
            throw std::runtime_error("initiator needs peerStatic + peerDeviceId (from the QR)");
        }
        hs_ = tether::noise::NoiseHandshake::initiator(opts_.staticKeypair, opts_.peerStatic);
        log("initiator: sending handshake msg1 -> " + shortId(remoteId_));
        signaling_->Relay(remoteId_, hs_->writeMessage());
    } else {
        hs_ = tether::noise::NoiseHandshake::responder(opts_.staticKeypair);
        log("responder: waiting for handshake msg1...");
    }
    return fut;
}

void SecureLink::send(const Bytes& message) {
    if (!transport_) throw std::runtime_error("not paired yet");
    signaling_->Relay(remoteId_, transport_->send.encryptWithAd({}, message));
}

void SecureLink::close() { signaling_->Close(); }

void SecureLink::onRegistered(int64_t) {
    log("registered as " + shortId(deviceId_));
    if (!registeredSettled_) {
        registeredSettled_ = true;
        registered_.set_value();
    }
}

void SecureLink::onError(const std::string& code, const std::string& message) {
    auto e = std::make_exception_ptr(std::runtime_error("broker error: " + code + " " + message));
    if (!registeredSettled_) {
        registeredSettled_ = true;
        registered_.set_exception(e);
    }
    if (!pairedSettled_) {
        pairedSettled_ = true;
        paired_.set_exception(e);
    }
}

void SecureLink::onDeliver(const std::string& from, const Bytes& payload) {
    // Transport phase.
    if (transport_) {
        if (opts_.onMessage) opts_.onMessage(transport_->recv.decryptWithAd({}, payload));
        return;
    }

    // Handshake phase.
    if (!hs_) return;
    if (opts_.role == Role::Responder && remoteId_.empty()) {
        remoteId_ = from;
        log("responder: received msg1 from " + shortId(from));
    }

    try {
        hs_->readMessage(payload);
        if (hs_->isMyTurn()) {
            log("replying with handshake msg2 -> " + shortId(remoteId_));
            signaling_->Relay(remoteId_, hs_->writeMessage());
        }
        if (hs_->isComplete()) {
            std::string authenticatedId = tether::identity::deviceIdFromPublicKey(hs_->remoteStaticKey());
            if (authenticatedId != remoteId_) {
                if (!pairedSettled_) {
                    pairedSettled_ = true;
                    paired_.set_exception(std::make_exception_ptr(std::runtime_error(
                        "identity mismatch: peer key fingerprints to " + shortId(authenticatedId) +
                        ", expected " + shortId(remoteId_))));
                }
                return;
            }
            transport_ = hs_->split();
            log("handshake complete, peer verified as " + shortId(authenticatedId));
            if (!pairedSettled_) {
                pairedSettled_ = true;
                paired_.set_value();
            }
        }
    } catch (...) {
        if (!pairedSettled_) {
            pairedSettled_ = true;
            paired_.set_exception(std::current_exception());
        }
    }
}

} // namespace tether::link

File: apps/android/app/src/main/kotlin/app/tether/net/SecureLink.kt
package app.tether.net

import app.tether.pairing.DeviceIdentity
import app.tether.pairing.NoiseHandshake
import app.tether.pairing.StaticKeypair
import app.tether.pairing.TransportPair
import kotlinx.coroutines.CompletableDeferred

/**
 * Client-side secure link — Kotlin port of apps/reference-cli/src/link.ts (SPEC §4).
 *
 * register → Noise_IK handshake over the broker's opaque relay → verify the
 * peer's authenticated static key fingerprints to the expected (scanned) device
 * id → encrypted transport. The broker never sees plaintext.
 *
 * Keep this behaviourally identical to the TS reference and the C++ port; the
 * shared vectors (docs/noise-test-vectors.json) pin the wire format.
 */
class SecureLink(
    serverUrl: String,
    private val staticKeypair: StaticKeypair,
    private val role: Role,
    private val peerStatic: ByteArray? = null,
    peerDeviceId: String? = null,
    private val onMessage: (ByteArray) -> Unit = {},
    private val logger: (String) -> Unit = {},
) : SignalingClient.Events {

    enum class Role { INITIATOR, RESPONDER }

    val deviceId: String = DeviceIdentity.deviceId(staticKeypair.publicKey)

    private val signaling = SignalingClient(serverUrl, deviceId, staticKeypair.publicKey, this)
    private var handshake: NoiseHandshake? = null
    private var transport: TransportPair? = null
    private var remoteId: String? = peerDeviceId

    private val registered = CompletableDeferred<Unit>()
    private val paired = CompletableDeferred<Unit>()

    /** Connect and register; suspends until the broker acknowledges. */
    suspend fun connect() {
        signaling.connect()
        registered.await()
    }

    /** Run the Noise_IK handshake; suspends until the transport is ready. */
    suspend fun pair() {
        when (role) {
            Role.INITIATOR -> {
                val ps = requireNotNull(peerStatic) { "initiator needs peerStatic (from the QR)" }
                remoteId = requireNotNull(remoteId) { "initiator needs peerDeviceId (from the QR)" }
                val hs = NoiseHandshake.initiator(staticKeypair, ps)
                handshake = hs
                logger("initiator: sending handshake msg1 -> ${short(remoteId!!)}")
                signaling.relay(remoteId!!, hs.writeMessage())
            }
            Role.RESPONDER -> {
                handshake = NoiseHandshake.responder(staticKeypair)
                logger("responder: waiting for handshake msg1…")
            }
        }
        paired.await()
    }

    fun send(message: ByteArray) {
        val t = checkNotNull(transport) { "not paired yet" }
        signaling.relay(remoteId!!, t.send.encryptWithAd(ByteArray(0), message))
    }

    fun send(message: String) = send(message.toByteArray())

    fun close() = signaling.close()

    // ---- SignalingClient.Events ----

    override fun onRegistered(heartbeatIntervalMs: Long) {
        logger("registered as ${short(deviceId)}")
        registered.complete(Unit)
    }

    override fun onPeerStatus(deviceId: String, online: Boolean) { /* presence UI hook */ }

    override fun onError(code: String, message: String) {
        val e = IllegalStateException("broker error: $code $message")
        if (!registered.isCompleted) registered.completeExceptionally(e)
        if (!paired.isCompleted) paired.completeExceptionally(e)
    }

    override fun onDeliver(from: String, payload: ByteArray) {
        // Transport phase: decrypt and surface to the caller.
        transport?.let {
            onMessage(it.recv.decryptWithAd(ByteArray(0), payload))
            return
        }

        // Handshake phase.
        val hs = handshake ?: return
        if (role == Role.RESPONDER && remoteId == null) {
            remoteId = from
            logger("responder: received msg1 from ${short(from)}")
        }

        try {
            hs.readMessage(payload)
            if (hs.isMyTurn) {
                logger("$role: replying with handshake msg2 -> ${short(remoteId!!)}")
                signaling.relay(remoteId!!, hs.writeMessage())
            }
            if (hs.isComplete) {
                val authenticatedId = DeviceIdentity.deviceId(hs.remoteStaticKey!!)
                if (authenticatedId != remoteId) {
                    paired.completeExceptionally(
                        IllegalStateException(
                            "identity mismatch: peer key fingerprints to ${short(authenticatedId)}, expected ${short(remoteId!!)}",
                        ),
                    )
                    return
                }
                transport = hs.split()
                logger("$role: handshake complete, peer verified as ${short(authenticatedId)}")
                paired.complete(Unit)
            }
        } catch (e: Exception) {
            if (!paired.isCompleted) paired.completeExceptionally(e)
        }
    }

    private fun short(id: String): String = id.take(8) + "…"
}
